test-browser/connection.spec.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright 2020 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Connection } from '../lib/esm/puppeteer/common/Connection.js';
+import expect from '../node_modules/expect/build-es5/index.js';
+
+/**
+ * A fake transport that echoes the message it got back and pretends to have got a result.
+ *
+ * In actual pptr code we expect that `result` is returned from the message
+ * being sent with some data, so we fake that in the `send` method.
+ *
+ * We don't define `onmessage` here because Puppeteer's Connection class will
+ * define an `onmessage` for us.
+ */
+class EchoTransport {
+  send(message) {
+    const object = JSON.parse(message);
+    const fakeMessageResult = {
+      result: `fake-test-result-${object.method}`,
+    };
+    this.onmessage(
+      JSON.stringify({
+        ...object,
+        ...fakeMessageResult,
+      })
+    );
+  }
+
+  close() {}
+}
+
+describe('Connection', () => {
+  it('can be created in the browser and send/receive messages', async () => {
+    let receivedOutput = '';
+    const connection = new Connection('fake-url', new EchoTransport());
+
+    /**
+     * 1 here is the ID because pptr increments a counter from 0 for each
```

Comment:
nit: That took me a couple of reads, just because you started with the number 1, I think!

Reply:
ah yeah this could be worded more clearly, thanks! I'll tweak.